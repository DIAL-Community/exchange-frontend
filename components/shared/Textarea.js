import React from 'react'
import classNames from 'classnames'

export const Textarea = React.forwardRef(({ value, placeholder, onChange, onBlur, isInvalid = false, className, ...otherProps }, ref) => (
  <textarea
    {...otherProps}
    ref={ref}
    value={value}
    onChange={onChange}
    onBlur={onBlur}
    placeholder={placeholder}
    className={classNames({ 'validation-error': isInvalid }, className, 'textarea w-full')}
  />
))

Textarea.displayName = 'Textarea'

export default Textarea
