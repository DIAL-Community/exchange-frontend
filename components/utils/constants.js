export const DisplayType = {
  LARGE_CARD: 'large-card',
  SMALL_CARD: 'small-card',
  PINNED_CARD: 'pinned-card',
  FEATURED_CARD: 'featured-card',
  HUB_CARD: 'hub-card',
  GRID_CARD: 'grid-card'
}

export const MainDisplayType = {
  LIST: 'list',
  GRID: 'grid'
}

export const ORIGIN_SLUG_EXPANSIONS = {
  'dial': 'Digital Impact Alliance',
  'digital-government-platform-tracker': 'Digital Government Platform Tracker',
  'digital-health-atlas': 'Digital Health Atlas',
  'digital-square': 'Digital Square',
  'dpga': 'Digital Public Goods Alliance',
  'giz': 'GIZ',
  'indiastack': 'India Stack',
  'manually-entered': 'Manually Entered',
  'unicef-covid': 'UNICEF'
}

export const OpportunityStatus = {
  CLOSED: 'CLOSED',
  OPEN: 'OPEN',
  UPCOMING: 'UPCOMING'
}

export const OpportunityType = {
  BID: 'BID',
  BUILDING_BLOCK: 'BUILDING BLOCK',
  INNOVATION: 'INNOVATION',
  TENDER: 'TENDER',
  OTHER: 'OTHER'
}

export const MaturityStatus = {
  DRAFT: 'DRAFT',
  PUBLISHED: 'PUBLISHED'
}

export const CategoryType = {
  DPI: 'DPI',
  FUNCTIONAL: 'FUNCTIONAL'
}

export const MappingStatus = {
  BETA: 'BETA',
  MATURE: 'MATURE',
  SELF_REPORTED: 'SELF-REPORTED',
  VALIDATED: 'VALIDATED'
}

export const LicenseTypeFilter = {
  ALL: 'all_license',
  COMMERCIAL: 'commercial_only',
  OPEN_SOURCE:'oss_only'
}

export const ObjectType = {
  BUILDING_BLOCK: 'BUILDING_BLOCK',
  DATASET: 'OPEN_DATA',
  PRODUCT: 'PRODUCT',
  PROJECT: 'PROJECT',
  OPPORTUNITY: 'OPPORTUNITY',
  ORGANIZATION: 'ORGANIZATION',
  USE_CASE: 'USE_CASE',
  WORKFLOW: 'WORKFLOW',
  MOVE: 'MOVE',
  PLAY: 'PLAY',
  PLAYBOOK: 'PLAYBOOK',
  RUBRIC_CATEGORY: 'RUBRIC_CATEGORY',
  CATEGORY_INDICATOR: 'CATEGORY_INDICATOR',
  CANDIDATE_DATASET: 'CANDIDATE_OPEN_DATA',
  CANDIDATE_ORGANIZATION: 'CANDIDATE_ORGANIZATION',
  CANDIDATE_PRODUCT: 'CANDIDATE_PRODUCT',
  CANDIDATE_RESOURCE: 'CANDIDATE_RESOURCE',
  CANDIDATE_ROLE: 'CANDIDATE_ROLE',
  TAG: 'TAG',
  SECTOR: 'SECTOR',
  URL: 'URL',
  CITY: 'CITY',
  COUNTRY: 'COUNTRY',
  REGION: 'REGION',
  TASK: 'TASK',
  USER: 'USER',
  CONTACT: 'CONTACT',
  RESOURCE: 'RESOURCE',
  RESOURCE_TOPIC: 'RESOURCE_TOPIC',
  SITE_SETTING: 'SITE_SETTING',
  TENANT_SETTING: 'TENANT_SETTING'

}

export const CategoryIndicatorType = {
  NUMERIC: 'numeric',
  SCALE: 'scale',
  BOOLEAN: 'boolean'
}

export const CandidateActionType = {
  APPROVE: 'APPROVE',
  REJECT: 'REJECT'
}

export const DEFAULT_PAGE_SIZE = 8

export const ProductStageType = {
  PILOT: 'pilot',
  SCALING: 'scaling',
  MATURE: 'mature'
}

export const ProductExtraAttributeNames = [
  'Relevance',
  'Impact',
  'Local Ownership',
  'Funders'
]

export const ProductMaturityAttributeNames = [
  'Deployments',
  'Deployed Countries',
  'Active Users',
  'Transactions per month',
  'Annual Revenue',
  'Funding Raised'
]
