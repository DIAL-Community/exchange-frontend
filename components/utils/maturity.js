export const getCategoryIndicatorTypes = (format) => [
  { label: format('shared.categoryIndicator.type.numeric'), value: 'numeric' },
  { label: format('shared.categoryIndicator.type.scale'), value: 'scale' },
  { label: format('shared.categoryIndicator.type.boolean'), value: 'boolean' }
]

export const getCategoryIndicatorScaleOptions = (format) => [
  { label: format('shared.categoryIndicator.type.scale.low'), value: 'low' },
  { label: format('shared.categoryIndicator.type.scale.medium'), value: 'medium' },
  { label: format('shared.categoryIndicator.type.scale.high'), value: 'high' }
]

export const getCategoryIndicatorBooleanOptions = (format) => [
  { label: format('shared.categoryIndicator.type.boolean.true'), value: 't' },
  { label: format('shared.categoryIndicator.type.boolean.false'), value: 'f' }
]

export const getCategoryIndicatorNumericOptions = () =>
  ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10'].map(value => ({ label: value, value }))
