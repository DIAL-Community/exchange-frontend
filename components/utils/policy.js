import { GRAPH_QUERY_CONTEXT_KEY, GRAPH_QUERY_POLICY_SLUG } from '../../lib/apolloClient'

const operationAllowed = (error, operationName) => {
  if (error?.graphQLErrors) {
    const { graphQLErrors } = error
    if (graphQLErrors.length > 0) {
      graphQLErrors.forEach(e => {
        if (e.extensions.operation === operationName) {
          return true
        }
      })
    }
  }

  return false
}

// Return map of operation and boolean value for each operation
// e.g. { editing: true, deleting: true }
export const fetchOperationPolicies = async (client, query, operations, variables = { slug: GRAPH_QUERY_POLICY_SLUG }) => {
  const response = await client.query({
    query,
    variables,
    errorPolicy: 'all',
    context: {
      headers: {
        [GRAPH_QUERY_CONTEXT_KEY]: operations.join(', ')
      }
    }
  })

  const policies = {}
  operations.forEach(operation => {
    policies[operation] = operationAllowed(response.error, operation)
  })

  return policies
}
