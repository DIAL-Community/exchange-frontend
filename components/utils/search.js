import { GRAPH_QUERY_CONTEXT } from '../../lib/apolloClient'

const MIN_SEARCH_VALUE_CHARACTERS = 2

export const fetchSelectOptions = async (client, input, query, fetchedDataCallback, locale = null) => {
  const variables = {
    search: input
  }

  if (input && input.trim().length < MIN_SEARCH_VALUE_CHARACTERS) {
    return []
  } else if (locale) {
    variables.locale = locale
  }

  const response = await client.query({
    query,
    variables,
    context: {
      headers: {
        ...GRAPH_QUERY_CONTEXT.VIEWING
      }
    }
  })

  return response.data ? fetchedDataCallback(response.data) : []
}

export const fetchSelectOptionsWithMaturity = async (client, input, query, fetchedDataCallback, locale = null) => {
  const variables = {
    mature: true,
    search: input
  }

  if (input && input.trim().length < MIN_SEARCH_VALUE_CHARACTERS) {
    return []
  } else if (locale) {
    variables.locale = locale
  }

  const response = await client.query({
    query,
    variables,
    context: {
      headers: {
        ...GRAPH_QUERY_CONTEXT.VIEWING
      }
    }
  })

  return response.data ? fetchedDataCallback(response.data) : []
}
