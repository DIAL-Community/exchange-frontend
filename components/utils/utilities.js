export const prependUrlWithProtocol = (url) => '//' + url

export const isValidFn = (maybeFunction) =>
  maybeFunction && {}.toString.call(maybeFunction) === '[object Function]'

export const isDebugLoggingEnabled = () => String(process.env.NEXT_PUBLIC_ENABLE_DEBUG).toLowerCase() === 'true'
