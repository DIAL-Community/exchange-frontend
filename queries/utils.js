const MIN_SEARCH_VALUE_CHARACTERS = 2

export const fetchSelectOptions = async (client, input, query, fetchedDataCallback, locale = null) => {
  const variables = {
    search: input
  }

  if (input && input.trim().length < MIN_SEARCH_VALUE_CHARACTERS) {
    return []
  } else if (locale) {
    variables.locale = locale
  }

  const response = await client.query({ query, variables })

  return response.data ? fetchedDataCallback(response.data) : []
}

export const fetchSelectOptionsWithMature = async (client, input, query, fetchedDataCallback, locale = null) => {
  const variables = {
    search: input,
    mature: true
  }

  if (input && input.trim().length < MIN_SEARCH_VALUE_CHARACTERS) {
    return []
  } else if (locale) {
    variables.locale = locale
  }

  const response = await client.query({ query, variables })

  return response.data ? fetchedDataCallback(response.data) : []
}
