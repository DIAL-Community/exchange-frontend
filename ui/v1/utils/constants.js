export const DisplayType = {
  LARGE_CARD: 'large-card',
  SMALL_CARD: 'small-card',
  PINNED_CARD: 'pinned-card'
}

export const OpportunityStatus = {
  CLOSED: 'CLOSED',
  OPEN: 'OPEN',
  UPCOMING: 'UPCOMING'
}

export const OpportunityType = {
  BID: 'BID',
  BUILDING_BLOCK: 'BUILDING BLOCK',
  INNOVATION: 'INNOVATION',
  TENDER: 'TENDER',
  OTHER: 'OTHER'
}

export const MaturityStatus = {
  DRAFT: 'DRAFT',
  PUBLISHED: 'PUBLISHED'
}

export const CategoryType = {
  DPI: 'DPI',
  FUNCTIONAL: 'FUNCTIONAL'
}

export const MappingStatus = {
  BETA: 'BETA',
  MATURE: 'MATURE',
  SELF_REPORTED: 'SELF-REPORTED',
  VALIDATED: 'VALIDATED'
}

export const LicenseTypeFilter = {
  ALL: 'all_license',
  COMMERCIAL: 'commercial_only',
  OPEN_SOURCE:'oss_only'
}

export const ObjectType = {
  BUILDING_BLOCK: 'BUILDING_BLOCK',
  DATASET: 'OPEN_DATA',
  PRODUCT: 'PRODUCT',
  PROJECT: 'PROJECT',
  OPPORTUNITY: 'OPPORTUNITY',
  ORGANIZATION: 'ORGANIZATION',
  PLAYBOOK: 'PLAYBOOK',
  USE_CASE: 'USE_CASE',
  CANDIDATE_DATASET: 'CANDIDATE_OPEN_DATA',
  CANDIDATE_ORGANIZATION: 'CANDIDATE_ORGANIZATION',
  CANDIDATE_PRODUCT: 'CANDIDATE_PRODUCT',
  CANDIDATE_ROLE: 'CANDIDATE_ROLE',
  TAG: 'TAG',
  SECTOR: 'SECTOR',
  COUNTRY: 'COUNTRY',
  CITY: 'CITY',
  CONTACT: 'CONTACT',
  URL: 'URL',
  PLAY: 'PLAY',
  MOVE: 'MOVE'
}

export const CategoryIndicatorType = {
  NUMERIC: 'numeric',
  SCALE: 'scale',
  BOOLEAN: 'boolean'
}

export const CandidateActionType = {
  APPROVE: 'APPROVE',
  REJECT: 'REJECT'
}

export const DEFAULT_PAGE_SIZE = 8
