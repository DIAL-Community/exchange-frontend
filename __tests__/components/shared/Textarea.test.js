import { render, fireEvent } from '@testing-library/react'
import Textarea from '../../../components/shared/Textarea'
import { mockNextUseRouter } from '../../utils/nextMockImplementation'

mockNextUseRouter()

describe('Unit test for the Textarea component.', () => {
  const TEST_ID = 'test-textarea'
  const mockOnChange = jest.fn()

  describe('Should match snapshot', () =>{
    test('valid.', () => {
      const { container } = render(<Textarea value='test value' placeholder='test placeholder' onChange={mockOnChange} />)
      expect(container).toMatchSnapshot()
    })

    test('invalid.', () => {
      const { container } = render(
        <Textarea value='test value' placeholder='test placeholder' onChange={mockOnChange} isInvalid />
      )
      expect(container).toMatchSnapshot()
    })

    test('10 rows high.', () => {
      const { container } = render(
        <Textarea value='test value' placeholder='test placeholder' onChange={mockOnChange} rows={10} />
      )
      expect(container).toMatchSnapshot()
    })
  })

  test('Should call onChange function.', () => {
    const input = render(<Textarea data-testid={TEST_ID} value='test' onChange={mockOnChange} />).getByTestId(TEST_ID)
    fireEvent.change(input, { target: { value: 'updated' } })
    expect(mockOnChange).toHaveBeenCalledTimes(1)
  })
})
