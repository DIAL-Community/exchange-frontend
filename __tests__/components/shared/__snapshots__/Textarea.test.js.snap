// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit test for the Textarea component. Should match snapshot 10 rows high. 1`] = `
<div>
  <textarea
    class="textarea w-full"
    placeholder="test placeholder"
    rows="10"
  >
    test value
  </textarea>
</div>
`;

exports[`Unit test for the Textarea component. Should match snapshot invalid. 1`] = `
<div>
  <textarea
    class="validation-error textarea w-full"
    placeholder="test placeholder"
  >
    test value
  </textarea>
</div>
`;

exports[`Unit test for the Textarea component. Should match snapshot valid. 1`] = `
<div>
  <textarea
    class="textarea w-full"
    placeholder="test placeholder"
  >
    test value
  </textarea>
</div>
`;
